Retrieve user profile on SET_AUTH_TOKEN 🔥

willdurand opened the pull request:
Refs #1820 #1644 #1751

---

This PR does a few things, but the main goal was to fetch the current logged user profile. I hooked into the `SET_AUTH_TOKEN` action to fetch the user profile with a saga. This action is cool because it is used in both SSR and not SSR.

### What has changed?

- The `userId` has moved to the `user` reducer now (it was in `api` since #2983).
- I removed the JWT token parsing to exclusively rely on the user profile data. This is now gone forever, but if we want to validate this token somehow, maybe we could re-introduce the method with a different name, _i.e._ `validateToken()`.

### What has been added?

- a `user` ducks that owns the current logged user profile
- a `user` saga that fetches the user profile on `SET_AUTH_TOKEN`
- a selector `isAuthenticated` in `user` reducer to know whether user is logged in

### What is missing? // TODO

- [x] write tests to cover the change in `core/server/base`. The `setAuthToken()` call to dispatch `SET_AUTH_TOKEN` was triggered too early and the saga could not catch it. I had to move the saga instantiation earlier. See this part of the code for further details: https://github.com/willdurand/addons-frontend/blob/master/src/core/server/base.js#L244-L260
- [x] how to handle error in the `user` saga? Since it is not created by UI, I don't know how to pass a error handler (id). Not sure if I have to actually..

kumar303 commented on tests/unit/core/server/testServer.js:
```diff
@@ -6,57 +6,74 @@ import { reducer as reduxAsyncConnect } from 'redux-connect';
-        ),
-        sagaMiddleware,
-      };
+      store = createStore(
```
I know it's convenient but it's too confusing to let `testClient()` have the side effect of creating a store in the outer scope. Let me know if you need ideas for how to fix it.

willdurand replied:
Then, I will pass the store to the `testClient()`

kumar303 commented on tests/unit/core/server/testServer.js:
```diff
@@ -6,57 +6,74 @@ import { reducer as reduxAsyncConnect } from 'redux-connect';
 
   describe('app', () => {
     it('varies on DNT', async () => {
-      const response = await testClient().get('/en-US/firefox/').end();
```
Unfortunately, `.end()` must always be the final call to the `testClient()` chain. This is because `end()` will shut down the server. So, without it, every one of these tests will be spawning a server that will never get shut down. It's the most terrible interface ever written. See https://github.com/visionmedia/supertest/issues/100

I originally couldn't think of a way to do this automatically but maybe we can just use a wrapper around `testClient()` ?

willdurand replied:
Oh ok 😢  

I did that to get rid of warnings saying that there was a "double callback" problem, see: https://github.com/visionmedia/supertest/issues/141.

kumar303 commented on tests/unit/core/server/testServer.js:
```diff
@@ -83,9 +100,66 @@ describe('core/server/base', () => {
       const response = await testClient({ stubRoutes })
-        .get('/en-US/firefox/simulation-of-a-non-existent-page').end();
+        .get('/en-US/firefox/simulation-of-a-non-existent-page')
+      ;
```
for consistency with the rest of the code base, the semi-colon should always go at the end of the line.

We'll add a rule 😇 https://github.com/mozilla/eslint-config-amo/issues/24

willdurand replied:
oups

willdurand commented on src/core/sagas/user.js:
```diff
@@ -0,0 +1,34 @@
+
+    yield put(userProfileLoaded({ profile: response }));
+  } catch (error) {
+    log.warn(`Current user profile failed to load: ${error}`);
```
perfect!

willdurand commented on src/core/sagas/user.js:
```diff
@@ -0,0 +1,34 @@
+}
+
+export default function* userSaga() {
+  yield takeLatest(SET_AUTH_TOKEN, fetchUserProfile);
```
We need the `username` in the header for instance, so I think we'll need to fetch the user profile anyway. I chose this action because it seems to be done only once on first hit (+ page reloads).

willdurand commented on src/core/reducers/user.js:
```diff
@@ -0,0 +1,33 @@
+  username: null,
+};
+
+export const userProfileLoaded = ({ profile }) => {
```
oh right!

willdurand commented on src/core/api/index.js:
```diff
@@ -327,3 +328,5 @@ export function autocomplete({ api, filters }: AutocompleteParams) {
   });
 }
+
+export const userProfile = userApi.userProfile;
```
heh I dont know, I saw @tofumatt doing something similar too.

willdurand commented:
I have updated the code.

tofumatt commented on src/core/server/base.js:
```diff
@@ -230,6 +230,7 @@ function baseServer(routes, createStore, { appSagas, appInstanceName = appName }
       let htmlLang;
       let locale;
       let pageProps;
+      let runningSagas;
```
Any reason this needs to be scoped up here? I know the scoping in this file is a bit broad and that's fine, just double-checking. The GitHub diff on this one is a bit harder to get context from.

willdurand replied:
That's because `store` and `sagaMiddleware` are there too..

tofumatt replied:
Cool, that's what I thought, makes sense! 👍

tofumatt commented on tests/unit/core/api/test_user.js:
```diff
@@ -0,0 +1,37 @@
+        .returns(mockResponse());
+
+      return api.userProfile({
+        api: { clientApp: CLIENT_APP_ANDROID, lang: 'en-US' },
```
This should use `dispatchClientMetadata().store.getState().api`. I know it's way more verbose/work but we should definitely get away from stubbing out raw state. We used to do it and it always leads to refactors or oddness.

tofumatt commented on tests/unit/core/containers/TestServerHtml.js:
```diff
@@ -12,14 +12,14 @@ import FakeApp, {
 } from 'tests/unit/core/server/fakeApp';
 
 describe('<ServerHtml />', () => {
-  const _helmentCanUseDOM = Helmet.canUseDOM;
```
lol Nice catch! 😆

tofumatt commented on tests/unit/core/reducers/test_user.js:
```diff
@@ -0,0 +1,42 @@
+  describe('reducer', () => {
+    it('initializes properly', () => {
+      const { id, username } = reducer(undefined);
+      expect(id).toBe(null);
```
I'd be fine if we just used `toEqual` everywhere because it's easier to read/rewrite/etc. I don't think we'll get bit by using it over `toBe`. @muffinresearch and I chatted about this and it's just nicer to have fewer matchers IMO.

That said if you left it here I'd be fine with it. Just a note.

willdurand replied:
can I `sed -i s/toBe/toEqual` in another PR maybe?

tofumatt replied:
Yeah, sure, it's really no biggie. I just figured I'd mention for future patches. I think `toEqual` is a nicer matcher too; `toBe` implies more of an exact type-check not a value-check.

tofumatt commented on tests/unit/core/reducers/test_user.js:
```diff
@@ -0,0 +1,42 @@
+    });
+
+    it('ignores unrelated actions', () => {
+      const state = { id: 12345, username: 'john' };
```
Could this use the `initialiState` from the actual reducer instead? I just like to stick with realistic data.

willdurand replied:
That would mean exposing the `initialState`?

tofumatt replied:
Yeah, [we do that in a fair number of reducers](https://github.com/mozilla/addons-frontend/blob/master/src/amo/reducers/featured.js#L4), I think it's fine.

tofumatt commented on src/core/sagas/user.js:
```diff
@@ -0,0 +1,29 @@
+
+  const state = yield select(getState);
+
+  const response = yield call(userProfileApi, {
```
I think this should have a `try/catch` around it... at least that's what we do for all of our other sagas and then we use `yield put(errorHandler.createErrorAction(error));` to dispatch an error.

See something like the search saga: https://github.com/mozilla/addons-frontend/blob/master/src/core/sagas/search.js#L21

willdurand replied:
Yeah I know how this works, but the UI gives the error handler ID, which is not possible here. That's why @kumar303 suggested to bubble up the errors: https://github.com/mozilla/addons-frontend/pull/2989#discussion_r134889238.

tofumatt replied:
Ahhhh, right. Okay, can you add a comment in here explaining why this saga is different than usual then and why we aren't using `errorHandler`? Sorry I missed some of the comments here.

willdurand replied:
Yes, just updated the code.

tofumatt commented on tests/unit/core/sagas/test_user.js:
```diff
@@ -0,0 +1,63 @@
+    expect(calledAction).toEqual(expectedCalledAction);
+  });
+
+  it('lets exceptions to be thrown', () => {
```
I think this test spec could be better written as `it('handles exceptions and dispatches an error', () => { [...] })`

But then as mentioned before I think the way this saga handles errors isn't how we tend to. I guess we should write some documentation at least on error handling, because we have a HOC and reducer for dispatching errors that components can consume.

kumar303 commented on tests/unit/core/sagas/test_user.js:
```diff
@@ -0,0 +1,63 @@
+    sagaTester.dispatch(setAuthToken(userAuthToken()));
+
+    expect.assertions(1);
+    return rootTask.done.catch((e) => {
```
I find `expect.assertions(N)` to be hard to maintain as apps grow. They also result in poor failure messages.

Here's a more robust way to do it:

```js
return rootTask.done
  .then(() => {
    throw new Error('unexpected success');
  })
  .catch((sagaError) => {
    mockApi.verify();
    expect(sagaError).toBe(error);
  });
```

btw, we try to avoid single character variable names in the code base.

kumar303 commented on tests/unit/core/server/testServer.js:
```diff
@@ -83,9 +111,68 @@ describe('core/server/base', () => {
       const response = await testClient({ stubRoutes })
-        .get('/en-US/firefox/simulation-of-a-non-existent-page').end();
+        .get('/en-US/firefox/simulation-of-a-non-existent-page')
+        .end();
```
I had an idea for making this better. What do you think of this?

```js
const response = await testClient({ stubRoutes })
  .get('/en-US/firefox/simulation-of-a-non-existent-page');
shudownServer(response);
```

`shudownServer()` would just call `response.end()` but it would make it really clear what's happening. Also, we can link to https://github.com/visionmedia/supertest/issues/100 in a comment.

tofumatt replied:
I dig it.

willdurand replied:
mmh this is not as easy as what I thought since `await` resolves the promise. I do not get a `supertest` instance but a response (with `req` and `res`) that is the same with or without `end()`. In other words, I cannot call `.end()` on `response`.

willdurand replied:
A solution could be to transform the `testClient()` function into something like `createApp()`, then have a call to `supertest(app)` to get the test client, and later shutdown the server:

```
const app = createApp(...);
const client = supertest(app);

// ...

app.close();
```

WDYT? Slighty more verbose, but explicit.

kumar303 commented on tests/unit/core/server/testServer.js:
```diff
@@ -83,9 +111,68 @@ describe('core/server/base', () => {
+      });
+      const response = await testClient({ store, sagaMiddleware })
+        .get('/en-US/firefox/')
+        .set('cookie', `${config.get('cookieName')}="${token}"`)
```
there's no `.cookie()` method?

willdurand replied:
I did not find anything. This PR seems to confirm this: https://github.com/visionmedia/supertest/pull/410.

willdurand replied:
I found a `cookie` method on the response only...

kumar303 commented on tests/unit/core/server/testServer.js:
```diff
@@ -83,9 +111,68 @@ describe('core/server/base', () => {
+        jsonData: createUserProfileResponse({ id: 42, username: 'babar' }),
+      });
+
+      mockWindow.expects('fetch')
```
I think it would be better just to check `mockUserApi.expects('userProfile')`

kumar303 commented on tests/unit/core/server/testServer.js:
```diff
@@ -23,32 +36,47 @@ describe('core/server/base', () => {
-    const app = baseServer(stubRoutes, createStoreAndSagas, {
-      appSagas: fakeSaga,
+    const app = baseServer(stubRoutes, _createStoreAndSagas, {
+      appSagas: appSagas || fakeSaga,
```
Can you also try to add test coverage for the `disableSSR=true` case? I think you could do it with this patch:

```js
diff --git a/src/core/server/base.js b/src/core/server/base.js
index e2f727a9..5927a744 100644
--- a/src/core/server/base.js
+++ b/src/core/server/base.js
@@ -3,7 +3,7 @@ import path from 'path';
 
 import 'babel-polyfill';
 import { oneLine } from 'common-tags';
-import config from 'config';
+import defaultConfig from 'config';
 import Express from 'express';
 import helmet from 'helmet';
 import Raven from 'raven';
@@ -116,7 +116,10 @@ function hydrateOnClient({ res, props = {}, pageProps }) {
     .end();
 }
 
-function baseServer(routes, createStore, { appSagas, appInstanceName = appName } = {}) {
+function baseServer(routes, createStore, {
+  appSagas, appInstanceName = appName,
+  config = defaultConfig,
+} = {}) {
   const app = new Express();
   app.disable('x-powered-by');
 
```

willdurand replied:
This works and I see the coverage in the report, but I haven't found a good assertion yet.

willdurand commented on tests/unit/core/reducers/test_user.js:
```diff
@@ -0,0 +1,51 @@
+    it('ignores unrelated actions', () => {
+      const state = {
+        ...initialState,
+        id: 12345,
```
But how can I be sure it does not change the values? The `initialState` has `null` values, which I think are not the best values for this test case.

willdurand replied:
I used an action creator to have a non-null state.

willdurand commented on tests/unit/core/sagas/test_user.js:
```diff
@@ -0,0 +1,63 @@
+
+  beforeEach(() => {
+    mockApi = sinon.mock(api);
+    const initialState = dispatchSignInActions().state;
```
of course!

willdurand commented:
I did not change the `.end()` since I dont know how to do it properly. We could maybe `listen()` to the `app` and therefore have full control over the server.. WDYT?

kumar303 commented:
> I did not change the .end() since I dont know how to do it properly.

Aw, whoops, I see what you mean about `await`. Thanks for giving it a try. We'll just have to remember to call `.end()` for now.

kumar303 commented on src/core/reducers/api.js:
```diff
@@ -48,24 +46,6 @@ export const initialApiState: ApiStateType = {
   userId: null,
```
`userId` should be removed from `initialApiState`. I think this will break some things. Those things will need to use `state.user.id` instead.

willdurand replied:
doh, bad rebase it seems because I am pretty sure I removed it :x

willdurand commented:
> There are a few problems:
> 
> First, the token is getting set but not the user object. Maybe the user fetching saga isn't running?

wow that's pretty bad. I think I have been tricked by my login state... I will double check and fix that.

> Secondly, some components are checking api.token and some are checking user.id to see if the user is logged in. All components should check user.id (I think?). We need to pick one.

heh yeah. The only thing I have done here is getting rid of `userId` in the `api` reducer (well, except that I introduced it again during a bad rebase...). I think #1644 mentions the use of `state.token` to determine whether a user is authenticated. I do agree with you we should pick one and I'd like to choose the `user.id`. Yet, I did not do it here because I did not want to increase the diff size.

Should I do the change here or open a new issue and send a new patch for the `user.id` vs `state.token`?

willdurand commented:
> I noticed a problem with disableSSR: true. Set that in your local config and restart your server. Go to an extension detail page and log in. Scroll down a bit. It should show you star ratings to select from but it does not:

mmh, I cannot reproduce this actually.

willdurand commented:
> wow that's pretty bad. I think I have been tricked by my login state... I will double check and fix that.

pff, I am silly I guess. The test case for that passes by chance. I was not 100% confident with the test case for that, because I did not find a good assertion 😪

willdurand commented:
Ok, I was able to reproduce the issue in the test suite. I fixed it and made sure we had assertions for it. I use [`cheerio`](https://cheerio.js.org/) because it is a Enzyme dependency.

willdurand commented on src/core/reducers/user.js:
```diff
@@ -1,14 +1,21 @@
 const USER_PROFILE_LOADED = 'USER_PROFILE_LOADED';
 
-export const initialState = {
+type Action = Object;
```
TODO: have a better `Action` type. I find it difficult and a bit superfluous to maintain this type in general.

willdurand commented:
Added a last commit with `state.token` changes.

kumar303 commented on tests/unit/core/server/testServer.js:
```diff
@@ -23,33 +33,50 @@ describe('core/server/base', () => {
-        sagaMiddleware,
-      };
+  function createStoreAndSagas({
+    reducers = { reduxAsyncConnect },
```
I think this should default to:

```js
{ reduxAsyncConnect, api: apiReducer, user: userReducer }
```

because really the server requires all of those to operate. I don't see a strong case for customizing the reducer list in any of the tests below.

willdurand replied:
great, it simplifies the test cases.

kumar303 commented on tests/unit/core/server/testServer.js:
```diff
@@ -83,9 +110,112 @@ describe('core/server/base', () => {
+      expect(api.token).toEqual(token);
+    });
+
+    it('runs sagas as early as possible', async () => {
```
The "early" part is an implementation detail. I think this spec could be "it fetches the user profile when given a token"

kumar303 commented on tests/unit/core/server/testServer.js:
```diff
@@ -83,9 +110,112 @@ describe('core/server/base', () => {
+      expect(user.username).toEqual('babar');
+      mockUserApi.verify();
+
+      // Parse the HTML response to retrieve the serialized redux state.
```
This is a great test! However, I think it should just be a standalone test. The spec could be "it serializes the redux state in html"

kumar303 commented on tests/unit/core/server/testServer.js:
```diff
@@ -83,9 +110,112 @@ describe('core/server/base', () => {
+      // We use `cloneDeep()` to allow modifications on the `config` object,
+      // since a call to `get()` makes it immutable. This is the case in the
+      // previous test cases (on `defaultConfig`).
+      const config = configUtil.cloneDeep(defaultConfig);
```
Interesting approach. I like it because it makes the config object realistic. It may be a little heavyweight since config objects are big but I'm not worried for this patch.

We actually have a few instances of config mocking that might benefit from this approach instead.

willdurand replied:
I was thinking about adding a helper function but could not find a similar snippet.

kumar303 commented on tests/unit/core/server/testServer.js:
```diff
@@ -83,9 +110,112 @@ describe('core/server/base', () => {
+      mockUserApi.verify();
+    });
+
+    it('sends the redux-saga END action when `disableSSR` is `true`', async () => {
```
This spec should be more related to the actual feature. I suggest this: "it fetches the user profile even when SSR is disabled"

kumar303 commented on tests/unit/core/server/testServer.js:
```diff
@@ -83,9 +110,112 @@ describe('core/server/base', () => {
+      mockUserApi
+        .expects('userProfile')
+        .once()
+        .returns(Promise.resolve(profile));
```
I would like to see another test that does `.returns(Promise.reject(new Error('example of an API error')))` and asserts a 500, like `expect(response.statusCode).toEqual(500);`

kumar303 commented on tests/unit/helpers.js:
```diff
@@ -276,3 +276,23 @@ export function createApiResponse({
+    num_addons_listed: 0,
+    occupation: '',
+    picture_type: '',
+    picture_url: 'https://addons-dev-cdn.allizom.org/static/img/zamboni/anon_user.png',
```
maybe this can use the `amoCDN` config property too?

kumar303 commented on src/core/api/user.js:
```diff
@@ -0,0 +1,12 @@
+import type { ApiStateType } from 'core/reducers/api';
+
+
+export function userProfile({ api }: {| api: ApiStateType |}) {
```
It would be pretty confusing if the caller forgot to pass a value for `api` so I think this should throw an error if it's empty.

willdurand replied:
can we open an issue for that? this should be done across all `core/api/*` functions.

willdurand replied:
can we open an issue for that? this should be done across all `core/api/*` functions.

kumar303 replied:
I don't think it should be done for all API functions. In this case, it's only necessary to require `api` because you have to be authenticated to make a request to the profile endpoint.

willdurand replied:
ok, adding the check here then

kumar303 commented on tests/unit/amo/helpers.js:
```diff
@@ -108,11 +110,15 @@ export function dispatchClientMetadata({
   const { store } = dispatchClientMetadata(otherArgs);
 
   store.dispatch(setAuthToken(authToken));
+  store.dispatch(userProfileLoaded({
```
oh yeah, good idea

kumar303 commented on src/core/containers/LoginRequired/index.js:
```diff
@@ -7,7 +7,7 @@ import LoginPage from 'core/components/LoginPage';
 export function mapStateToProps(state) {
   return {
-    authenticated: !!state.api.token,
+    authenticated: !!state.user.id,
```
It looks like you didn't have to update any tests for this which is a red flag. Can you try to add test coverage for it?

willdurand replied:
It is actually covered by the change in `dispatchSignInActions()` 😉

kumar303 replied:
oh I see, that's cool. The helper is doing its job well.

You could edit the specs in `TestLoginRequired` but it's not a big deal. I would propose the following changes:

* old: it sets authenticated to true when there is a token
  * new: it sets authenticated to true
* old: it sets authenticated to false when there is not a token
  * new: it sets authenticated to false

willdurand commented on src/core/server/base.js:
```diff
@@ -36,25 +36,25 @@ import I18nProvider from 'core/i18n/Provider';
-const appName = config.get('appName');
+const appName = defaultConfig.get('appName');
 
 export function getPageProps({ noScriptStyles = '', store, req, res }) {
```
Yeah, that definitely makes sense.

willdurand commented on src/core/reducers/user.js:
```diff
@@ -0,0 +1,47 @@
+};
+
+export const initialState: UserStateType = {
+  id: null,
```
All have been replaced. `api.token` is used for the API calls, but not for the authentication status checks anymore. I think having a new prop makes sense. Another option would be a selector:

```
// in core/reducers/user
const isAuthenticated = (state) => {
    return !!state.user.id;
};
```

The benefit is to not have to update the code if we want to change internal state things, and the state does not get bigger.

kumar303 replied:
Sure. A selector sounds good.

willdurand commented:
Ok I think this is ready. I had to pull your patch from #3023 because test suite was failing on Travis and I wanted to make sure it was not because of my changes. I'll get rid of this commit before merging this PR (or rebase when #3023 gets merged).

willdurand commented:
Ok I think this is ready.

@kumar303 I had to pull your patch from #3023 because test suite was failing on Travis and I wanted to make sure it was not because of my changes. I'll get rid of this commit before merging this PR (or rebase when #3023 gets merged).

willdurand commented:
I have updated the code, and also changed the action type/creator I introduced to match the naming pattern as per https://github.com/mozilla/addons-frontend/pull/3025#discussion_r135909932.

kumar303 commented on src/core/reducers/user.js:
```diff
@@ -0,0 +1,51 @@
+
+export default function reducer(
+  state: UserStateType = initialState,
+  action: Action = {}
```
for future reference, this would look more like:

```js
action: LoadUserProfileAction | LogOutUserAction
```

...but we'd have to switch away from constants to strings so that Flow [can figure out which action is which](https://flow.org/en/docs/react/redux/#toc-typing-redux-reducers) in the switch.

kumar303 commented on src/core/server/base.js:
```diff
@@ -333,27 +339,29 @@ function baseServer(routes, createStore, { appSagas, appInstanceName = appName }
 }
 
 export function runServer({
-  listen = true, app = appName, exitProcess = true,
```
Hmm, I can't see where any callers were using this `app` parameter. I guess it's safe to remove.

kumar303 commented on tests/unit/core/components/TestAuthenticateButton.js:
```diff
@@ -65,31 +70,32 @@ describe('<AuthenticateButton />', () => {
-      auth: {},
-      api: { lang: 'en-GB' },
-    });
+    const startLoginUrlStub = sinon.stub(api, 'startLoginUrl').returns('https://a.m.org/login');
```
just FYI, the reason our lines tend to be under 80 chars is because I like to use three splits in Vim. I haven't been able to convince people that our 100 line limit is too long for that :)

willdurand replied:
ah I only have two splits in Vim 😅 

I think that 80 chars is not enough, but I am sold! Let's go for 80 chars.

kumar303 commented on tests/unit/core/containers/TestLoginRequired.js:
```diff
@@ -41,15 +42,13 @@ describe('<LoginRequired />', () => {
     });
 
-    it('sets authenticated to false when there is not a token', () => {
+    it('it sets authenticated to false', () => {
```
technically the `it` isn't needed but Jest doesn't prefix our specs with "it" like mocha did so they aren't readable anyway, heh.

willdurand replied:
oh good catch, that was actually a mistake 😞

kumar303 commented on tests/unit/core/sagas/test_user.js:
```diff
@@ -0,0 +1,66 @@
+      .then(() => {
+        throw new Error('unexpected success');
+      })
+      .catch((expectedError) => {
```
We tend to only use *expected* for variables set by the tests, not variables returned by the program under test. No need to bother with changing it, just though I'd mention it.

kumar303 commented on tests/unit/core/server/testServer.js:
```diff
@@ -83,9 +110,151 @@ describe('core/server/base', () => {
+      // since a call to `get()` makes it immutable. This is the case in the
+      // previous test cases (on `defaultConfig`).
+      const config = configUtil.cloneDeep(defaultConfig);
+      config.disableSSR = true;
```
I don't think it's so important to test this with `disableSSR=true`. It would be better not to just because it will simulate production better.

willdurand replied:
I see what you mean, and I will add another similar test for that.

I'd like to keep this one though as it actually covers the fact that with `disableSSR=true`, sagas are executed correctly and both user profile and token are passed to the client side. The bug you mentioned earlier this week was because of that.

willdurand replied:
The state computed on the server is still passed to the client _via_ the HTML. `SET_AUTH_TOKEN` is not often called on the client because it is already done by the server. That is also why I put these assertions in the `disableSSR=true` test case. That is the HTML I get with `disableSSR=true`:

``` html
<!DOCTYPE html>
<html lang="en-US" dir="ltr" data-reactroot="" data-reactid="1" data-react-checksum="1242348688">
 <head data-reactid="2">
  <meta charset="utf-8" data-reactid="3" />
  <meta name="viewport" content="width=device-width, initial-scale=1" data-reactid="4" />
  <link rel="shortcut icon" href="/favicon.ico" data-reactid="5" />
  <title data-react-helmet="true" data-reactid="6"></title>
  <noscript data-reactid="7">
   <style data-reactid="8">/* !important is needed to show the button, make sure switch is hidden */
.InstallButton .InstallButton-button {
  display: inline-block !important;
}

.InstallButton .InstallButton-switch {
  display: none !important;
}
</style>
  </noscript>
 </head>
 <body data-reactid="9">
  <div id="react-view" data-reactid="10"></div>
  <script type="application/json" id="redux-store-state" data-reactid="11">{"addons":{},"api":{"clientApp":"firefox","lang":"en-US","token":"eyJhdXRoX2hhc2giOiJiNTRhNTBiYjYzMDg0OTVkMzQ3OTU5MGIyMDg4MmEyZDkwZWM3ZmJjIiwidXNlcl9pZCI6MTA2MzkxNTZ9:1dmdwc:DHZayyST3Y2Ms6SE7memWNz5l7Y","userAgent":"Mozilla\u002F5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit\u002F537.36 (KHTML, like Gecko) Chrome\u002F59.0.3071.115 Safari\u002F537.36","userAgentInfo":{"browser":{"name":"Chrome","version":"59.0.3071.115","major":"59"},"os":{"name":"Mac OS","version":"10.12.6"}}},"autocomplete":{"loading":false,"suggestions":[]},"categories":{"categories":null,"loading":false},"errors":{},"errorPage":{"clearOnNext":false,"error":null,"hasError":false,"statusCode":null},"featured":{"addonType":null,"loading":false,"results":[]},"infoDialog":{},"installations":{},"landing":{"addonType":null,"featured":{"count":0,"results":[]},"highlyRated":{"count":0,"results":[]},"loading":false,"popular":{"count":0,"results":[]},"resultsLoaded":false},"loadingBar":0,"reduxAsyncConnect":{"loaded":false,"loadState":{}},"reviews":{"byAddon":{}},"search":{"count":0,"filters":{},"loading":false,"results":[]},"user":{"id":10639156,"username":"willdurand"},"viewContext":{"context":"VIEW_CONTEXT_EXPLORE"}}</script>
  <script src="//127.0.0.1:3001/amo-c623c6e9e4951a016e4b.js" data-reactid="12"></script>
 </body>
</html>
```

I have updated both test cases now. The `disableSSR=true` one checks that the state is correct (so that login works with such a configuration). The second test case about making sure the state is serialized into HTML now simulates the production configuration.

kumar303 replied:
> I'd like to keep this one though as it actually covers the fact that with disableSSR=true, sagas are executed correctly and both user profile and token are passed to the client side.

Adding both test cases is fine but I didn't explain myself very well in my suggestion to only include a test for the `disableSSR: false` case (to emulate production). Here is some more context.

The `core/server/testServer.js` suite performs high level integration tests--that is, at the server->client level. These types of tests are good but they are also dangerous because they tempt you to test for a matrix of possibilities and that matrix can be vast. The more test cases you add, the slower the suite becomes.

I think the best approach for an app like ours is to rely on unit tests as much as possible and then try to use static analysis (like Flow) to make sure our units are glued together right. We may not ever get to that last part because I haven't been able to convince the team of how Flow can accomplish this for us.

So, we should still do integration level tests but we should use them sparingly. For this case, we actually [already have a unit test](https://github.com/mozilla/addons-frontend/blob/master/tests/unit/core/containers/TestServerHtml.js#L101) for rendering the redux state to the client. We were lacking an integration test for it so that's why I requested adding one (it could definitely catch a bug or two) but I don't think we need to try and cover every edge case.

willdurand replied:
Oh, but you are absolutely right! Thanks for the explanations.

willdurand commented:
@tofumatt I am going to fix the last @kumar303's comments before merging it. Would you have something to raise or could you unlock your review status?
